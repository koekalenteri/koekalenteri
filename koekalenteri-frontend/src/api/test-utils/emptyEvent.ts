import type { Event } from 'koekalenteri-shared/model';

export const emptyEvent: Event = {
  accountNumber: 'test',
  allowHandlerMembershipPriority: true,
  allowOwnerMembershipPriority: true,
  classes: [{ class: 'test' }],
  cost: 123,
  costMember: 123,
  createdAt: new Date('1989-01-01'),
  createdBy: 'test',
  description: 'test',
  endDate: new Date('1990-01-01'),
  entries: 0,
  entryEndDate: new Date('1990-01-01'),
  entryStartDate: new Date('1990-01-01'),
  eventType: 'test',
  id: 'test',
  judges: [123],
  location: 'test',
  modifiedAt: new Date('1989-01-02'),
  modifiedBy: 'test',
  name: 'test',
  official: {
    id: 1,
    name: 'Teemu Toimitsija',
    email: 'joo@ei.com',
    phone: '040-official',
    location: 'Suomi',
    district: 'Helsinki',
    eventTypes: ['test']
  },
  secretary: {
    id: 2,
    name: 'Siiri Sihteeri',
    email: 'ei@joo.com',
    phone: '040-secretary',
    location: 'Ruotsi'
  },
  organizer: {
    id: 1,
    name: 'Suomen Noutajakoirajärjestö ry'
  },
  paymentDetails: 'test',
  places: 10,
  referenceNumber: 'test',
  startDate: new Date('1990-01-01'),
  state: 'confirmed'
}
