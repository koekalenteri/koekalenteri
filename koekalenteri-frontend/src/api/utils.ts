import type { ConfirmedEventEx, Event, EventEx, JsonEvent } from 'koekalenteri-shared/model';
import { endOfDay, startOfDay, subDays } from 'date-fns';
import { DEFAULT_EVENT } from './defaultEvent';

// https://stackoverflow.com/a/69756175/10359775
type PickByType<T, Value> = {
  [P in keyof T as T[P] extends Value ? P : never]: T[P]
}
type EventDates = keyof PickByType<Event, Date|undefined>;

const EVENT_DATE_PROPS: EventDates[] = ['startDate', 'endDate', 'entryStartDate', 'entryEndDate', 'createdAt', 'modifiedAt'];

export function toDate(value: string | undefined): Date | undefined {
  return value ? new Date(value) : undefined;
}

function rehydrateDate(value: string | number | Date | undefined) {
  if (value instanceof Date) {
    return value;
  }
  if (value) {
    return new Date(value);
  }
}

export function rehydrateEvent(event: Partial<JsonEvent> | Partial<Event>, now = new Date()): EventEx {
  for (const prop of EVENT_DATE_PROPS) {
    event[prop] = rehydrateDate(event[prop]);
  }
  if (event.deletedAt) {
    event.deletedAt = new Date(event.deletedAt);
  }

  for (const cls of event.classes || []) {
    if (typeof cls === 'string') {
      continue;
    }
    cls.date = rehydrateDate(cls.date || event.startDate);
  }

  let isEntryOpen = false;
  let isEntryClosing = false;
  let isEntryUpcoming = false;
  let isEntryClosed = false;
  let isEventUpcoming = false;
  let isEventOngoing = false;
  let isEventOver = false;

  if (event.state === 'confirmed') {
    const confirmedEvent = event as ConfirmedEventEx;
    isEventUpcoming = confirmedEvent.startDate > now;
    isEntryUpcoming = confirmedEvent.entryStartDate > now;
    isEntryOpen = startOfDay(confirmedEvent.entryStartDate) <= now &&
      endOfDay(confirmedEvent.entryEndDate) >= now;
    isEntryClosing = isEntryOpen && subDays(confirmedEvent.entryEndDate, 7) <= endOfDay(now);
    isEntryClosed = endOfDay(confirmedEvent.entryEndDate) < now &&
      startOfDay(confirmedEvent.startDate) > now;
    isEventOngoing = confirmedEvent.startDate <= now && confirmedEvent.endDate >= now;
    isEventOver = confirmedEvent.endDate < now;
  }

  let statusText: 'tentative' | 'cancelled' | 'extended' | undefined;
  if (event.state === 'tentative' || event.state === 'cancelled') {
    statusText = event.state;
  }
  if (event.entryOrigEndDate) {
    statusText = 'extended';
  }

  return {
    ...DEFAULT_EVENT,
    ...event as Partial<Event>,
    isEntryUpcoming,
    isEntryOpen,
    isEntryClosing,
    isEntryClosed,
    isEventUpcoming,
    isEventOngoing,
    isEventOver,
    statusText
  };
}
