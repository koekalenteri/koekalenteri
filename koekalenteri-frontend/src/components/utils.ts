import { isValid, isSameMonth, lightFormat, isSameDay, parseISO, isSameYear } from "date-fns";

export function dateSpan(start: Date | string, end: Date | string) {
  if (typeof start === 'string') {
    start = parseISO(start);
  }
  if (typeof end === 'string') {
    end = parseISO(end) || start;
  }
  if (!isValid(start)) {
    return '';
  }
  end = end || start;
  if (isSameDay(start, end)) {
    return lightFormat(start, 'd.M.yyyy');
  }
  if (isSameMonth(start, end)) {
    return lightFormat(start, 'd') + '-' + lightFormat(end, 'd.M.yyyy');
  }
  if (isSameYear(start, end)) {
    return lightFormat(start, 'd.M') + '-' + lightFormat(end, 'd.M.yyyy');
  }
  return lightFormat(start, 'd.M.yyyy') + ' - ' + lightFormat(end, 'd.M.yyyy');
}
