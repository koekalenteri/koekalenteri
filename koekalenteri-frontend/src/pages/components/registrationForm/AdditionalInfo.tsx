import { ChangeEventHandler, useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { TextField } from '@mui/material'
import { Registration } from 'koekalenteri-shared/model'

import useDebouncedCallback from '../../../hooks/useDebouncedCallback'
import CollapsibleSection from '../CollapsibleSection'

type AdditionalInfoProps = {
  notes?: string
  onChange?: (props: Partial<Registration>) => void
  onOpenChange?: (value: boolean) => void
  open?: boolean
}

export function AdditionalInfo({ notes, onChange, onOpenChange, open }: AdditionalInfoProps) {
  const { t } = useTranslation()
  const [value, setValue] = useState(notes ?? '')

  const dispatchChange = useDebouncedCallback((notes: string) => onChange?.({ notes }))

  const handleChange = useCallback<ChangeEventHandler<HTMLTextAreaElement>>(
    (e) => {
      const newValue = e.target.value
      setValue(newValue)
      dispatchChange(newValue)
    },
    [dispatchChange]
  )

  useEffect(() => {
    setValue(notes ?? '')
  }, [notes])

  return (
    <CollapsibleSection title={t('registration.notes')} open={open} onOpenChange={onOpenChange}>
      <TextField
        label={t('registration.notes')}
        multiline
        onChange={handleChange}
        rows={4}
        sx={{ width: '100%' }}
        value={value}
      />
    </CollapsibleSection>
  )
}
