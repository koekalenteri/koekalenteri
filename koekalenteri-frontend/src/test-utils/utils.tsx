import { useEffect } from "react"
import { createMemoryRouter, createRoutesFromElements, RouteObject, RouterProvider } from "react-router-dom"
import { RouterInit } from '@remix-run/router'
import { act } from "@testing-library/react"
import mediaQuery from 'css-mediaquery'
import { RecoilValue, useRecoilValue } from "recoil"

/**
 * Abstraction to avoid re-writing all tests for the time being
 * @see https://github.com/remix-run/react-router/blob/main/packages/react-router/__tests__/data-memory-router-test.tsx
*/
export function DataMemoryRouter({
  basename,
  children,
  fallbackElement,
  hydrationData,
  initialEntries,
  initialIndex,
  routes,
}: {
  basename?: RouterInit["basename"];
  children?: React.ReactNode | React.ReactNode[];
  fallbackElement?: React.ReactNode;
  hydrationData?: RouterInit["hydrationData"];
  initialEntries?: string[];
  initialIndex?: number;
  routes?: RouteObject[];
}) {
  const router = createMemoryRouter(routes || createRoutesFromElements(children), {
    basename,
    hydrationData,
    initialEntries,
    initialIndex,
  })
  return <RouterProvider router={router} fallbackElement={fallbackElement} />
}

export function flushPromisesAndTimers(): Promise<void> {
  return act(
    () =>
      new Promise<void>(resolve => {
        setTimeout(resolve, 100)
        jest.runAllTimers()
      }),
  )
}

export function waitForDebounce() {
  return act(() => new Promise(resolve => setTimeout(resolve, 150)))
}

export const createMatchMedia = (width: number) => (query: string): MediaQueryList => ({
  matches: mediaQuery.match(query, { width }),
  media: query,
  onchange: null,
  addListener: jest.fn(), // deprecated
  removeListener: jest.fn(), // deprecated
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
  dispatchEvent: jest.fn(),
})

export function RecoilObserver<T>({node, onChange}: {node: RecoilValue<T>, onChange: (value: T) => void}) {
  const value = useRecoilValue(node)
  useEffect(() => onChange(value), [onChange, value])
  return null
}
