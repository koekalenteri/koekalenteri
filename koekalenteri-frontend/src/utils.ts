import type { EventEx } from 'koekalenteri-shared/model'

export function entryDateColor(event: EventEx) {
  if (!event.isEntryOpen) {
    return 'text.primary'
  }
  return event.isEntryClosing ? 'warning.main' : 'success.main'
}

export function unique(arr: string[]): string[] {
  return arr.filter((c, i, a) => a.indexOf(c) === i)
}

export function uniqueDate(arr: Date[]): Date[] {
  return arr.filter((c, i, a) => a.findIndex(f => f.valueOf() === c.valueOf()) === i)
}
