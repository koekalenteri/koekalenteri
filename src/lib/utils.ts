import type {
  DeepPartial,
  EventState,
  JsonValue,
  PublicContactInfo,
  PublicDogEvent,
  RegistrationDate,
  RegistrationTime,
} from '../types'

import { eachDayOfInterval, subDays } from 'date-fns'
import { diff } from 'deep-object-diff'

import { zonedEndOfDay, zonedStartOfDay } from '../i18n/dates'

type EventVitals = Partial<Pick<PublicDogEvent, 'startDate' | 'endDate' | 'entryStartDate' | 'entryEndDate' | 'state'>>

export const isValidForEntry = (state?: EventState) => !['draft', 'tentative', 'cancelled'].includes(state ?? '')

export const isEventUpcoming = ({ startDate }: EventVitals, now = new Date()) => !!startDate && startDate > now

export const isEntryUpcoming = ({ entryStartDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate && entryStartDate > now && (isValidForEntry(state) || state === 'tentative')

export const isEntryOpen = ({ entryStartDate, entryEndDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate &&
  !!entryEndDate &&
  zonedStartOfDay(entryStartDate) <= zonedEndOfDay(now) &&
  zonedEndOfDay(entryEndDate) >= zonedEndOfDay(now) &&
  isValidForEntry(state)

export const isEntryClosing = (event: EventVitals, now = new Date()) =>
  !!event.entryEndDate &&
  isEntryOpen(event, now) &&
  subDays(event.entryEndDate, 7) <= zonedEndOfDay(now) &&
  isValidForEntry(event.state)

export const isEntryClosed = ({ startDate, entryEndDate }: EventVitals, now = new Date()) =>
  !!startDate && !!entryEndDate && zonedEndOfDay(entryEndDate) < now && zonedStartOfDay(startDate) > now

export const isEventOngoing = ({ startDate, endDate, state }: EventVitals, now = new Date()) =>
  !!startDate &&
  !!endDate &&
  zonedStartOfDay(startDate) <= zonedEndOfDay(now) &&
  zonedEndOfDay(endDate) >= zonedEndOfDay(now) &&
  isValidForEntry(state) &&
  state !== 'confirmed'

export const isEventOver = ({ endDate }: EventVitals, now = new Date()) => !!endDate && zonedEndOfDay(endDate) < now

export const eventDates = (event?: Pick<PublicDogEvent, 'classes' | 'startDate' | 'endDate'> | null) => {
  if (!event) return []
  return event.classes.length
    ? uniqueDate(event.classes.map((c) => c.date ?? event.startDate))
    : eachDayOfInterval({ start: event.startDate, end: event.endDate })
}

export const uniqueClasses = (event?: Pick<PublicDogEvent, 'classes'> | null) =>
  unique((event?.classes ?? []).map((c) => c.class))

export const placesForClass = (
  event: DeepPartial<Pick<PublicDogEvent, 'places' | 'classes'>> | undefined | null,
  cls: string
) => {
  if (!event) {
    return 0
  }

  return (
    (event.classes ?? []).filter((c) => c.class === cls).reduce((acc, cur) => acc + (Number(cur.places) || 0), 0) ||
    Number(event.places) ||
    0
  )
}

export const uniqueClassDates = (event: PublicDogEvent, cls: string) =>
  cls === event.eventType
    ? eventDates(event)
    : uniqueDate(event.classes.filter((c) => c.class === cls).map((c) => c.date ?? event.startDate))

export const registrationDates = (event: PublicDogEvent, times: RegistrationTime[], cls?: string | null) =>
  (cls ? uniqueClassDates(event, cls) : eventDates(event)).flatMap<RegistrationDate>((date) =>
    times.map((time) => ({ date, time }))
  )

export const unique = <T = string>(arr: T[]): T[] => arr.filter((c, i, a) => a.indexOf(c) === i)

export const uniqueFn = <T>(arr: T[], cmp: (a: T, b: T) => boolean): T[] =>
  arr.filter((c, i, a) => a.findIndex((f) => cmp(f, c)) === i)

export const uniqueDate = (arr: Date[]) => [...new Set<number>(arr.map((d) => d.valueOf()))].map((v) => new Date(v))

const DATE_RE = /^\d{4}-(?:0[1-9]|1[0-2])-(?:[0-2][1-9]|[1-3]0|3[01])$/
const TIME_RE = /^(?:[0-1]\d|2[0-3])(?::[0-6]\d)(?::[0-6]\d)?(?:\.\d{3})?(?:[+-][0-2]\d:[0-5]\d|Z)?$/

export const isDateString = (value: unknown): value is string => {
  if (typeof value !== 'string') {
    return false
  }
  const [date, time] = value.split('T')
  return DATE_RE.test(date) && TIME_RE.test(time)
}

function dateReviver(_key: string, value: JsonValue): JsonValue | Date {
  if (isDateString(value)) {
    const dateObj = new Date(value)
    if (!isNaN(+dateObj)) {
      return dateObj
    }
  }
  return value
}

export const parseJSON = (json: string, reviveDates: boolean = true) => {
  if (!json) return undefined

  const reviver = reviveDates ? dateReviver : undefined

  return JSON.parse(json, reviver)
}

export type AnyObject = Record<string, any>
export type EmptyObject = Record<string, never>
export type Entries<T> = {
  [K in keyof T]: [K, T[K]]
}[keyof T][]

export const isEmpty = (o: AnyObject): o is EmptyObject => Object.keys(o).length === 0

export const isObject = (o: unknown): o is AnyObject =>
  o !== null && Object.prototype.toString.call(o) === '[object Object]'

export const isEmptyObject = (o: unknown): o is EmptyObject => isObject(o) && isEmpty(o)

export const hasChanges = (a: object | undefined | null, b: object | undefined | null): boolean =>
  !isEmptyObject(diff(a ?? {}, b ?? {}))

export const clone = <T extends AnyObject>(a: T): T => ({ ...a })

export const merge = <T extends AnyObject>(a: T, b: DeepPartial<T>): T => {
  const result = isObject(a) ? clone(a) : ({} as T)
  if (!isObject(b)) {
    return result
  }
  for (const [key, value] of Object.entries(b) as Entries<T>) {
    if (isObject(value)) {
      const old = result[key]
      result[key] = isObject(old) ? merge(old, value) : value
    } else {
      result[key] = value
    }
  }
  return result
}

// This function merges a patch object into a base object, but only creates new objects for paths that actually change
export const patchMerge = <T extends AnyObject, P extends DeepPartial<T>>(base: T, patch: P): T & P => {
  if (!isObject(base) || !isObject(patch)) return patch as T & P

  let changed = false
  const result = { ...base } as T & P
  for (const key of Object.keys(patch)) {
    const merged = patchMerge(base[key], (patch as any)[key])
    if (merged !== base[key]) {
      ;(result as any)[key] = merged
      changed = true
    }
  }

  return changed ? result : (base as T & P)
}

export const printContactInfo = (info?: PublicContactInfo) =>
  [info?.name, info?.phone, info?.email].filter(Boolean).join(', ')
