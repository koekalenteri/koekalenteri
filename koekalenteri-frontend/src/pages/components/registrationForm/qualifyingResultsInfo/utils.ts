import type { ManualTestResult, QualifyingResult } from 'koekalenteri-shared/model'
import type { EventResultRequirement, EventResultRequirements, EventResultRequirementsByDate } from '../../../../rules'

import { nanoid } from 'nanoid'

import { unique } from '../../../../utils'
import { objectContains } from '../validation'

const asArray = (v: EventResultRequirements | EventResultRequirement) => (Array.isArray(v) ? v : [v])

function findFirstMissing(requirements: EventResultRequirementsByDate | undefined, results: QualifyingResult[]) {
  if (!requirements) {
    return []
  }
  for (const rule of requirements.rules) {
    for (const opt of asArray(rule)) {
      const { count, ...rest } = opt
      if (results.filter((r) => objectContains(r, rest)).length < count) {
        return rest
      }
    }
  }
}

export function availableTypes(requirements?: EventResultRequirementsByDate) {
  if (!requirements) {
    return []
  }
  return unique(requirements.rules.flatMap((rule) => asArray(rule).map((opt) => opt.type)))
}

export function availableResults(requirements?: EventResultRequirementsByDate, type?: string) {
  if (!requirements) {
    return []
  }
  return unique(
    requirements.rules.flatMap((rule) =>
      asArray(rule)
        .filter((opt) => !type || opt.type === type)
        .map((opt) => (opt.cert ? 'CERT' : opt.result))
    )
  )
}

export function createMissingResult(
  requirements: EventResultRequirementsByDate | undefined,
  results: ManualTestResult[],
  regNo: string
): ManualTestResult {
  const rule = findFirstMissing(requirements, results)
  return {
    id: nanoid(10),
    regNo,
    date: new Date(),
    official: false,
    qualifying: true,
    type: '',
    judge: '',
    location: '',
    result: '',
    class: '',
    ...rule,
  }
}

export function resultBorderColor(qualifying: boolean | undefined) {
  if (qualifying === true) {
    return 'success.light'
  }
  if (qualifying === false) {
    return 'error.main'
  }
}

export function getResultId(result: ManualTestResult | QualifyingResult) {
  if ('id' in result) {
    return result.id
  }
  return result.type + result.class + result.date.toISOString()
}
