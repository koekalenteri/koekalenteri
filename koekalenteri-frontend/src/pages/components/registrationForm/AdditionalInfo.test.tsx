import { render, screen } from '@testing-library/react'

import { renderWithUserEvents, waitForDebounce } from '../../../test-utils/utils'

import { AdditionalInfo } from './AdditionalInfo'

describe('PlacesInput', () => {
  it('should render with minimal info', () => {
    const { container } = render(<AdditionalInfo />)
    expect(container).toMatchSnapshot()
  })

  it('should render with text', () => {
    const { container } = render(<AdditionalInfo notes="test" />)
    expect(container).toMatchSnapshot()
  })

  it('should rerender with new value', () => {
    const { container, rerender } = render(<AdditionalInfo notes="test a" />)
    rerender(<AdditionalInfo notes="changed notes" />)
    expect(container).toMatchSnapshot()
  })

  it('should call onChange', async () => {
    const onChange = jest.fn()
    const { user } = renderWithUserEvents(<AdditionalInfo notes="test" onChange={onChange} />)

    await waitForDebounce()
    expect(onChange).not.toHaveBeenCalled()

    const input = screen.getByRole('textbox')

    await user.clear(input)
    await waitForDebounce()
    expect(onChange).toHaveBeenLastCalledWith({ notes: '' })

    onChange.mockReset()

    await user.type(input, 'testing notes')
    await waitForDebounce()
    expect(onChange).toHaveBeenLastCalledWith({ notes: 'testing notes' })
  })
})
