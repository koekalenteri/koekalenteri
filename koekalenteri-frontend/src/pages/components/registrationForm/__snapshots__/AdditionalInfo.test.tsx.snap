// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesInput should render with minimal info 1`] = `
<div>
  <div
    class="MuiBox-root css-z0stvt"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-1qbrvqz"
    >
      <div
        class="MuiBox-root css-yxbw19"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          registration.notes
        </p>
        <p
          class="MuiFormHelperText-root css-gorocw-MuiFormHelperText-root"
        />
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiBox-root css-n2y79c"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for=":r0:"
                  id=":r0:-label"
                >
                  registration.notes
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r0:"
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  />
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        registration.notes
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PlacesInput should render with text 1`] = `
<div>
  <div
    class="MuiBox-root css-z0stvt"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-1qbrvqz"
    >
      <div
        class="MuiBox-root css-yxbw19"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          registration.notes
        </p>
        <p
          class="MuiFormHelperText-root css-gorocw-MuiFormHelperText-root"
        />
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiBox-root css-n2y79c"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for=":r1:"
                  id=":r1:-label"
                >
                  registration.notes
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r1:"
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  >
                    test
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        registration.notes
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PlacesInput should rerender with new value 1`] = `
<div>
  <div
    class="MuiBox-root css-z0stvt"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiBox-root css-1qbrvqz"
    >
      <div
        class="MuiBox-root css-yxbw19"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          registration.notes
        </p>
        <p
          class="MuiFormHelperText-root css-gorocw-MuiFormHelperText-root"
        />
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiBox-root css-n2y79c"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for=":r2:"
                  id=":r2:-label"
                >
                  registration.notes
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r2:"
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  >
                    changed notes
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        registration.notes
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
